js/image/shared/Webcam.svelte

```diff
@@ -33,24 +35,40 @@
 	}>();
 
 	onMount(() => (canvas = document.createElement("canvas")));
-	const size = {
-		width: { ideal: 1920 },
-		height: { ideal: 1440 }
+
+	const handle_device_change = async (event: InputEvent): Promise<void> => {
+		const target = event.target as HTMLInputElement;
+		const device_id = target.value;
+
+		await get_video_stream(include_audio, video_source, device_id).then(
+			async (local_stream) => {
+				stream = local_stream;
+				selected_device =
+					available_video_devices.find(
+						(device) => device.deviceId === device_id
+					) || null;
+				options_open = false;
+			}
+		);
 	};
-	async function access_webcam(device_id?: string): Promise<void> {
-		if (!navigator.mediaDevices || !navigator.mediaDevices.getUserMedia) {
-			dispatch("error", i18n("image.no_webcam_support"));
-			return;
-		}
+
+	async function access_webcam(): Promise<void> {
 		try {
-			stream = await navigator.mediaDevices.getUserMedia({
-				video: device_id ? { deviceId: { exact: device_id }, ...size } : size,
-				audio: include_audio
-			});
-			video_source.srcObject = stream;
-			video_source.muted = true;
-			video_source.play();
-			webcam_accessed = true;
+			get_video_stream(include_audio, video_source)
+				.then(async (local_stream) => {
+					webcam_accessed = true;
+					available_video_devices = await get_devices();
+					stream = local_stream;
+				})
+				.then(() => set_available_devices(available_video_devices))
+				.then((devices) => {
+					available_video_devices = devices;
+					selected_device = devices[0];
```

Comment:
```js
```js
selected_device =
  devices.find(
    (device) => device.label === stream.getVideoTracks()[0].label
  ) || available_video_devices[0];
```

Reply:
Or this as you found: https://stackoverflow.com/questions/66048083/enumeratedevices-after-getusermedia-how-to-find-the-active-devices/66072112#66072112

Reply:
Yeah, this is better:

```js
const used_devices = stream
  .getTracks()
  .map((track) => track.getSettings()?.deviceId)[0];

selected_device = used_devices
  ? devices.find(
      (device) => device.deviceId === (used_devices as string)
    ) || available_video_devices[0]
  : available_video_devices[0];
```